### DISCUSSION TITLE
Mimalloc

### DISCUSSION THREAD

**Discussion by ghost** [vlang/v#11861] [Ideas]
Mimalloc -> https://github.com/microsoft/mimalloc is a high performance allocator with superior performance to the default libc allocator, as well as jemalloc and other allocators. Implementing it as the default allocator for V could lead to significant performance gains.

**Comment by Cons-Cat** [comment]
Imho it might be better for V to implement a similar allocator written in V in the long run, so that it's portable to the non-C backends.

**Comment by dumblob** [comment]
I'd actually say this is totally out of scope of *V* (at least out of scope for *V 1.0*). `libc` on a given system is never perfect, but I'm still convinced that it's still much closer to what the user intends to use than to ignore it and use something own. And for `-freestanding`/bare (embedded systems etc.) use cases mimalloc is most probably not what the people want.

That being said if e.g. FreeBSD libc knew that its `malloc()` is slow, it'd most probably already bundle `mimalloc`. But it doesn't - and *V* can not say "hey, I'm the expert here and know better than all the FreeBSD devs how to do malloc() on your system and use mimalloc() instead".

On the other hand, `mimalloc` is cool (is it the default allocator on Windows 10+?) and e.g. on Linux it works much better with thread local storage than many other allocators and because *V* wants to leverage TLS a lot for its future parallelism, it might make sense to potentially give it a go...

> Imho it might be better for V to implement a similar allocator written in V in the long run, so that it's portable to the non-C backends.

I'm afraid this would be futile. First, it's a bigger piece of SW. Second it's actively maintained with a lot of manpower (and thus the *V* version would be a lot of work). Third it doesn't make much sense because `mimalloc` would be useful only with C backend anyway and only on a very strict subset of platforms *V* generally supports.

**Comment by Cons-Cat** [comment]
> is it the default allocator on Windows 10+?)

I think this document strongly implies that it is not.
https://microsoft.github.io/mimalloc/overrides.html
